openraft/src/membership/membership.rs

```diff
@@ -19,51 +19,113 @@ use crate::NodeId;
 /// every config.
 #[derive(Clone, Default, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Membership {
+    /// learners set
+    learners: BTreeSet<NodeId>,
+
     /// Multi configs.
     configs: Vec<BTreeSet<NodeId>>,
 
     /// Cache of all node ids.
-    all_nodes: BTreeSet<NodeId>,
+    all_members: BTreeSet<NodeId>,
 }
 
 impl MessageSummary for Membership {
     fn summary(&self) -> String {
-        let mut res = vec!["[".to_string()];
+        let mut res = vec!["members:[".to_string()];
         for (i, c) in self.configs.iter().enumerate() {
             if i > 0 {
                 res.push(",".to_string());
             }
             res.push(format!("{:?}", c));
         }
         res.push("]".to_string());
+
+        res.push(",learners:[".to_string());
+        for (learner_cnt, learner_id) in self.learners.iter().enumerate() {
+            if learner_cnt > 0 {
+                res.push(",".to_string());
+            }
+            res.push(format!("{:?}", learner_id));
+        }
+        res.push("]".to_string());
         res.join("")
     }
 }
 
 impl Membership {
     pub fn new_single(members: BTreeSet<NodeId>) -> Self {
         let configs = vec![members];
-        let all_nodes = Self::build_all_nodes(&configs);
-        Membership { configs, all_nodes }
+        let all_members = Self::build_all_members(&configs);
+        let learners = BTreeSet::new();
+        Membership {
+            learners,
+            configs,
+            all_members,
+        }
+    }
+
+    pub fn new_single_with_learners(members: BTreeSet<NodeId>, learners: BTreeSet<NodeId>) -> Self {
+        let configs = vec![members];
+        let all_members = Self::build_all_members(&configs);
+        Membership {
+            learners,
+            configs,
+            all_members,
+        }
```

Comment:
There should be tests to ensure that  `replace()`, `push()`, `new_single_with_learners()` and `new_multi_with_learners()` won't store a `node_id` in both `configs` and `learners`